suite/sanity/shrink.go

```diff
@@ -0,0 +1,47 @@
+package sanity
+
+import (
+	"github.com/gravitational/robotest/infra/gravity"
+
+	"github.com/sirupsen/logrus"
+)
+
+func shrink(p interface{}) (gravity.TestFunc, error) {
```

Comment:
So I'm curious about the UX of this. We already have a "resize" scenario which AFAIU is (was) supposed to mean both "expand" and "shrink". So if you resize 1->3, this means "expand", while e.g. 3->2 is "shrink".

I wonder if you have you considered augmenting the resize scenario to support node removal (as it should I think) instead of adding a separate "shrink" case?

Reply:
"resize" claimed to shrink, but it does not (currently) do so in practice. `recover` implements shrink and expand with several configurable parameters, and is in a state of disrepair. I considered repairing either or both of these as the way forward with shrink.  It would not be difficult to transition this work over to either.

To address "Why did you chose a new test case?"

As I see them, Robotest's largest pain points are:

1. determinism (developers don't trust results)
2. speed (it takes a long time for PR builds to compete)
3. development/contribution complexity (no one has the priority/wants to touch it because it is difficult and error prone to do so).

Implementing configurable test cases with structural branching tends to hurt us in 1 & 3.  That being said, we'll always have a balance within robotest (e.g. writing a separate test case for a different distro or number of nodes in the cluster feels foolish).  Currently, I see our dividing line as:

Items that affect the initial gravity cluster configuration (the neutral state at the start of the scenario we want to exercise) should be parameterized.  Items that determine what exactly is exercised during the scenario should be coded into the test code.  Following this out, I argue both `resize` and `recover` should be refactored into smaller independent test cases instead of expanded to cover more scenarios. `resize` -> `expand` and `shink`. `recover` into several independent `recoverX` (e.g. `recoverMaster`, `recoverKnode`) for the different scenarios that are of interest.

For some 3rd party thought on the situation I recommend some of Google's TOTT:

https://testing.googleblog.com/2018/06/testing-on-toilet-keep-tests-focused.html
https://testing.googleblog.com/2014/04/testing-on-toilet-test-behaviors-not.html

P.S. Getting off on a tangent, much of the provisioning that is currently implemented inline in tests should be factored out into fixtures -- which will provide further distinction between "something in the cloud went wrong" and "the behavior we were trying to test failed".  This data could then feed into metrics about utility: What is the ratio of ERR to FAIL?  Given our $$ estimates per scenario, where are we best served by investment? Furthermore, with provisioning factored out of test cases, the parameterization would (almost?) entirely be for the provisioner, and test code would simply receive pointers to the nodes the need, ready to go.  These ideas are non-trivial refactors of robotests internal harness... so I'm not going to further delay shrink or other near-term objectives while I explore them.

Reply:
Okay, sounds good. I was also thinking that we're currently lacking an ability to easily combine all our individual test-cases into more complicated scenarios, our current scenarios are pretty rigid. I.e. as implemented, "shrink" scenario also tests join even though we also have a separate scenario for it as well.

I think that ideally all operations we support should be sort of "building blocks" that robotest understands and then we can construct test scenarios easily in declarative way.

Regarding "resize" vs "expand"/"shrink" - it's probably just a question of semantics, but I think that it'd be nice to try and be consistent, i.e. if we have "shrink" then we probably also should have "expand" (esp. since "resize" really is basically expand). We don't need to do it in this PR though.

Reply:
> I.e. as implemented, "shrink" scenario also tests join even though we also have a separate scenario for it as well.

The join-before-shrink is 100% because I'm trying to regression test gravitational/gravity#1445. My initial shrink implementation did not have a join after the install (which failed to catch 1445 because it is dependent on the extra `knode` join).

I considered implementing this as two test cases, one as a pure shrink and a 2nd as a regression test specific to 1445.  This approach is "cleaner" test wise.  However, we're already running pretty hot in terms of robotest cloud spend::bugs caught, so I was reluctant to have two test cases when one got us most of the value.  This argument is a lot weaker since 1445 doesn't consistently repro with the current test case.  I'm still trying to figure out why -- but I wanted to get this up for review.

> I think that ideally all operations we support should be sort of "building blocks" that robotest understands and then we can construct test scenarios easily in declarative way.

As long as we keep the test cases in the robotest repo, our building blocks are `infra/*` (should probably be renamed lib at some point) and our test cases are `suite/sanity/*`.  As is, `suite/sanity/*` are not declarative, but they're getting closer.

If, however, we move the test cases into Gravity, having robotest be a library of building blocks would make a ton of sense. There is considerable merit having the test case definition live in Gravity, most notably: it becomes easier for gravity developers to contribute new test cases., and test cases merge/revert/whatever with other Gravity code. If we pursue this, we'll almost certainly want to library-ize robotest and find a way to have Gravity's test definitions in go.  We currently do bash -> json(-ish) config -> go -> terraform .  We can cut out the bash & json portion if we turn robotest into a library.  I don't think adding further complexity our bash & json-ish DSL is a sustainable future. 

Following this to its conclusion: Does it make sense to have robotest code as a separate repo?  Probably, but it may start to be treated more like planet.